src/Language/PureScript/AST/Declarations.hs

```diff
@@ -313,17 +313,67 @@ data DeclarationRef
   | ReExportRef SourceSpan ExportSource DeclarationRef
   deriving (Show, Generic, NFData, Serialise)
 
+-- |
+-- A version of DeclarationRef for easily derived comparisons.
+-- This version just has SourceSpan removed from all types.
+--
+data ComparableDeclarationRef
```

Comment:
Why do we need to create this additional data type? Note that comparing now allocates when it didn't before.

Reply:
I was hoping there would be some compiler magic that spares these allocations and treat this new "Comparable" data type more like a lens(?) into the original type.
The low-level behavior is still mysterious to me.
It's possible to write an `Ord` instance for `DeclarationRef` but it will be kinda verbose.

Reply:
We already do this the verbose way with `Type` in `Language.PureScript.Types`. Following the same pattern here would be preferable to me too.

Reply:
Will do.
Is there a reason why the `compareType` function is used? Couldn't these just all be renamed to `compare` and placed directly after the instance?
https://github.com/purescript/purescript/blob/9cad73ed8ea7df3011032ddbd2f3de5a0c08629c/src/Language/PureScript/Types.hs#L557-L558
https://github.com/purescript/purescript/blob/9cad73ed8ea7df3011032ddbd2f3de5a0c08629c/src/Language/PureScript/Types.hs#L578-L580

Reply:
Having a separate `compareType` function allows us to specify different type variables for the annotations for each argument, which means that we can encode the requirement that annotations shouldn't be considered in the types; this way the compiler can verify for us that we aren't accidentally comparing annotations. It also gives us a little more flexibility if we ever want to compare two `Type` values which do actually have different annotations.

Reply:
I guess that's less relevant in this instance since `DeclarationRef` doesn't take type parameters.